packages/ember-data/lib/serializers/json-serializer.js

```diff
@@ -710,17 +710,27 @@ export default Serializer.extend({
     @method normalizeUsingDeclaredMapping
     @private
   */
-  normalizeUsingDeclaredMapping: function(typeClass, hash) {
+  normalizeUsingDeclaredMapping: function(modelClass, hash) {
     var attrs = get(this, 'attrs');
-    var payloadKey, key;
+    var payloadKey, normalizedKey, key;
 
     if (attrs) {
       for (key in attrs) {
-        payloadKey = this._getMappedKey(key);
+        payloadKey = this._getMappedKey(key, modelClass);
+
         if (!hash.hasOwnProperty(payloadKey)) { continue; }
 
-        if (payloadKey !== key) {
-          hash[key] = hash[payloadKey];
+        // we need to normalize for JSONAPISerializer
```

Comment:
I think this comment should be removed as this is not really `JSONAPISerializer` specific. You would encounter this issue with any serializer where you customize `keyForAttribute`/`keyForRelationship` to not just pass through.